### ISSUE TITLE
Error writing to state database: IO error: ... Too many open files

### ISSUE THREAD

**Issue by yi-ji** [ethereum/cpp-ethereum#4493]
looking for a little help here... 
`eth` crushed after a few minutes running, saying:
`Error writing to state database: IO error: /.../.ethereum/d4e56740/12041/state/014656.ldb: Too many open files`

After a few times of retrying, `eth` quitted
`Fail writing to state database. Bombing out.`

Has anyone met this before?
I saw some one met this problem with go-ethereum but their discussion are not helping

some info:
eth version 1.3.0
eth network protocol version: 63
Client database version: 12041
`ulimit -n` returns 1024 which is normal

**Comment by gumb0** [comment]
I now see this sometimes, too

**Comment by gumb0** [comment]
Could be related https://github.com/google/leveldb/issues/401

**Comment by pirapira** [comment]
I saw this on leveldb 1.18.  Trying https://github.com/google/leveldb/commit/ea0a7586b8615fd39c6b8f5a8a21a1f242129c2f .

**Comment by gumb0** [comment]
go-ethereum seems to have had similar issues
https://github.com/ethereum/go-ethereum/issues/2314
https://github.com/ethereum/go-ethereum/issues/2358
https://github.com/ethereum/go-ethereum/issues/880
https://github.com/ethereum/go-ethereum/pull/929

**Comment by yi-ji** [comment]
Just my experience here:

I changed to `geth` and it said my system time needs to be sync correctly. After I did that, no more same errors for both `geth` and `eth`, although it seems irrelevant.

**Comment by pirapira** [comment]
That's a great insight.

**Comment by gumb0** [comment]
There are 18818 files in the state database on my ropsten node, no idea if this is ok

**Comment by pirapira** [comment]
I have a similar number.  They are at most 2.1 MB, so the number of files seems to grow when the database gets bigger.

**Comment by gumb0** [comment]
geth's problems are probably not related because goleveldb is completely separate implementation in go, not the wrapper over C++.

I suspect that this fix is what we need https://github.com/google/leveldb/commit/646c3588de84ac532a0e3525eae03edae1ea759f

**Comment by gumb0** [comment]
@chfast So is there an issue why we can't use newer leveldb?

**Comment by chfast** [comment]
No issue. Someone just have to do the hard work of packaging it for Hunter... I'm suggesting using this fork https://github.com/bitcoin-core/leveldb -- they at least have Windows support in place.

**Comment by gumb0** [comment]
Is there an easy way to just try it locally first? Like installing it as system-level leveldb and rebuilding eth?

**Comment by chfast** [comment]
1. Build it locally.
2. Pass hints to CMake where leveldb is. Usually, pass `-DLEVELDB_INCLUDE_DIR=... -DLEVELDB_LIBRARY=...`.  Check out FindLevelDB.cmake for this.

**Comment by gumb0** [comment]
Ok, built with bitcoin-core/leveldb v1.20, it gives me
```
 Error writing to state database: Corruption: corrupted compressed block contents
```
(even after killing database)

Will try next with google/leveldb

**Comment by chfast** [comment]
Ok, great you managed to test it.

**Comment by gumb0** [comment]
Yeah, it turns out it does work, it's just `-K` flag seems to be not enough to kill all previous database. After I killed `.ethereum` dir manually, it started to work, but then soon again gave me "too many open files" error, weird. Then I restarted it and it works fine for a long time already 😕 

Will continue to run it, also will try with the main net.

**Comment by gumb0** [comment]
@yi-ji or anyone else, you could try it, too, if you wish

What I did:
```
git clone https://github.com/bitcoin-core/leveldb
cd leveldb
make
cd ~/Ethereum/cpp-ethereum/build
cmake .. -DLEVELDB_INCLUDE_DIR=~/Ethereum/leveldb/include/ -DLEVELDB_LIBRARY=~/Ethereum/leveldb/out-static/
make eth
```

After that kill the data directory and restart sync.

**Comment by chfast** [comment]
Great! I will test it on the mainnet next week.

**Comment by gumb0** [comment]
Still happens on the main net with new LevelDB, though much less often

**Comment by gumb0** [comment]
Now I see it happening on Ropsten, too
